### PR TITLE
Upstream chunked volumes into bareos-16.2

### PR DESCRIPTION
**Pull request by mvwieringen**
This upstreams all patches needed to bareos-16.2 to support chunked volumes using the droplet
cloud library. Its on top of PR 86 which upstreams the stable fixes needed which need to merge
first. The second batch is including the renaming of the storage backend driver from object to droplet
naming including some fixes to the build environment.

It seems GitHub doesn't allow to have stacked PRs and people are ranting about that for ages.

This comes from 2 separate branches in the upstream repository.

"DON'T MERGE until pull request 86 is merged".

### CODE REVIEW

**Comment by joergsteffens** [comment]
Hello Marco,

thank you for your contribution. As you have seen, we integrated your pull request #61 into bareos-17.2. I see, you integrated 2 commits from me. Thank you for this. Now I try to figure out, 
  * is this on top of #61 or does it replace it?
  * you say, that pull request #67 must be integrated before. However, here I see, that branch mvw/stable -fixes is also integrated to this commit. Has this be done intentionally?
  * why have you splittet the pull request in the first place? Because #67 are just fixes for existing bug and #68 is adding new functionality?

I'll continue by integrating you pull request into our internal git and perform same tests on them.

regards,
Jörg

**Comment by joergsteffens** [comment]
I forgot to ask what version of https://github.com/mvwieringen/Droplet you use together with this and there which patchess are required to https://github.com/bareos/Droplet?

**Comment by mvwieringen** [comment]
> is this on top of #61 or does it replace it?

I closed pull request 61 after 4 months as nothing happened and spliced my repo to have
two branches e.g. mvw/stable-fixes which contains all patches for 16.2 which are needed to make
gluster and ceph usable. The second branch is indeed on top of mvw/stable-fixes and is called mvw/chunked-volumes and need the stable fixes. So essentially these 2 pull request replace #61.
The closing of #61 happened due to the fact that I removed the old branch and pushing these
two new branches.

With heavy stress testing we discovered more problems in the chunked volumes so there are more
patches in both branches divided like the previous ones into stable fixes (like typo fixes found during
debugging) and extra Chunked Volume (S3) fixes.

You probably have to do some conflict resolution when you merge this into 16.2 and later into 17.2
but then again git is pretty smart like already noticing if some patches are already applied (like the
renaming of source files, after which it will patch the right file in the branch that things got renamed to
even when the old branch still does patches to the old name of the files.)

> you say, that pull request #67 must be integrated before. However, here I see, that branch 
mvw/stable-fixes is also integrated to this commit. Has this be done intentionally?

Yes one branch is on top of the other only GitHub Pull request don't handle two branches very
cleanly. The whole idea was to give people a stable branch with patches and S3 as additional
addon without forcing to use experimental code which has big database changes. Honestly you
can talk for ages about integrating S3 but the stable patches belong to 16.2 and should have been
integrated as merge into there and merged later on into 17.2 and master.

> why have you splittet the pull request in the first place? Because #67 are just fixes for existing bug and #68 is adding new functionality?

Spot on. I like to have clear boundaries on the PRs. It only seems GitHub handles them pretty badly
but google has enough people complaining about that.

The whole idea this time is to integrate everything into 16.2 as that is what my and your customer
wants e.g. to stay on 16.2 as the whole stress testing has been done on 16.2 for over a year so
switching to some bleeding edge code like the just tagged 17.2 release is not an option for them.

> I forgot to ask what version of https://github.com/mvwieringen/Droplet you use together with this and there which patchess are required to https://github.com/bareos/Droplet?

This repo has one special branch which contains all patches needed and is on top of the last upstream
master branch of Scality. Most of it was once tried to get upstreamed but also there I gave up after 4 months of getting no response. The original bareos repo for Droplet had some hacked support for proxy support (which only partly works) and I once started on Path based S3 support but that is a lot
more work due to the V4 signature you have to calculate. I once had v2 signatures working but as
Amazon is forcing mostly v4 signature its of no value. I think it makes most sense to just drop the
old bareos-master branch as its unmaintained and contains some nasty bugs and have done stuff
which only works partly.

When you look at my Droplet repo it will show the correct branch by default same as for the Bareos
repo.

**Comment by joergsteffens** [comment]
Thank you for your explanation. Just to be sure: https://github.com/mvwieringen/Droplet/tree/mvw/upstream (https://github.com/mvwieringen/Droplet/commit/4e3f45cc8b664b6ba879272f8813df7d7ce7486d) is exactly the version you tested in your environment? Do you consider giving it a tag?

**Comment by mvwieringen** [comment]
> Thank you for your explanation. Just to be sure: https://github.com/mvwieringen/Droplet/tree/mvw/upstream (mvwieringen/Droplet@4e3f45c) is exactly the version you tested in your environment? Do you consider giving it a tag?

Yes that is the version that is running at the customer with self compiled binaries. Which has had
multiple Gb of data (maybe even 100 or more by now with all stress testing.)

The orginal idea was to upstream but Scality seems to have dropped any interest in it giving
the fact they have not updated since February. Like I said before I would merge the branch
into the bareos one and drop the current stuff. I think you can even create a PR yourself
from my branch to yours.

File: autoconf/configure.in [hunk 1]
```diff
@@ -4545,30 +4545,27 @@ AC_SUBST_FILE(DEBIAN_CONTROL_DIRECTOR_PYTHON_PLUGIN)
 dnl build a list of storage backends we need to build.
 BUILD_SD_BACKENDS=""
 if test x$use_libtool != xno; then
-   BUILD_SD_BACKENDS="${BUILD_SD_BACKENDS} libbareossd-fifo.la"
-   BUILD_SD_BACKENDS="${BUILD_SD_BACKENDS} libbareossd-gentape.la"
-   BUILD_SD_BACKENDS="${BUILD_SD_BACKENDS} libbareossd-tape.la"
-
-   if test X"$have_glusterfs" = "Xyes" ; then
-      BUILD_SD_BACKENDS="${BUILD_SD_BACKENDS} libbareossd-gfapi.la"
```
Comment (joergsteffens) [inline autoconf/configure.in]: libbareossd-gfapi.la got lost due to the cleanup here, causing build to fail. A fix is not required, as I already added it with an additional commit.

Comment (joergsteffens) [reply autoconf/configure.in]: I've been a bit optimistic on this. Compile still fails. I'll check for the reason later on.

Comment (mvwieringen) [reply autoconf/configure.in]: Strange thing compiled fine on my Linux install and it also compiled the stuff on the same platform which was used to build the customer rpms before. So could be some strange thing on OBS.

Comment (joergsteffens) [reply autoconf/configure.in]: Well, adding this AND regenerate "configure" (I missed that on the first try) solved the problem. So it does not seem to be OBS related.

**Comment by joergsteffens** [comment]
We integrated this PR in a separate branch: https://github.com/bareos/bareos/commits/bareos-16.2-droplet
We will integrate it to bareos-17.2 soon.

**Comment by joergsteffens** [comment]
This PR is part of Bareos (bareos-16.2-droplet, bareos-17.2 and master). As it will not be merged into bareos-16.2, this PR is not automatically closed. Therefore I close it now manually.

**Pull request closed without merging**